### PR TITLE
PRT-1082 stateless rpc consumer with cache

### PR DESCRIPTION
**Pull request by ranlavanet**

### CODE REVIEW

File: ecosystem/cache/handlers.go [hunk 1]
```diff
@@ -28,7 +29,9 @@ var (
 )
 
 const (
-	SEP = ";"
+	SharedStateWriteAttempts = 5
+	Int64Cost                = 8 // the cost is directly set to 8, which is the size in bytes of the int64
```
Comment (ranlavanet) [inline ecosystem/cache/handlers.go:33]: @omerlavanet  should we use this as the cost for the latest block set?

File: ecosystem/cache/handlers.go [hunk 2]
```diff
@@ -185,22 +202,37 @@ func (s *RelayerCacheServer) getLatestBlockInner(chainID string, providerAddr st
...
 	if existingLatest <= latestBlock { // equal refreshes latest if it expired
 		// we are setting this with a futuristic invalidation time, we still want the entry in cache to protect us from putting a lower last block
 		cacheStore := LastestCacheStore{latestBlock: latestBlock, latestExpirationTime: time.Now().Add(DefaultExpirationForNonFinalized)}
-		utils.LavaFormatDebug("setting latest block", utils.Attribute{Key: "providerAddr", Value: providerAddr}, utils.Attribute{Key: "chainID", Value: chainID}, utils.Attribute{Key: "latestBlock", Value: latestBlock})
-		s.CacheServer.finalizedCache.Set(latestBlockKey(chainID, providerAddr), cacheStore, cacheStore.Cost()) // no expiration time
+		utils.LavaFormatDebug("setting latest block", utils.Attribute{Key: "key", Value: key}, utils.Attribute{Key: "latestBlock", Value: latestBlock})
+		s.CacheServer.finalizedCache.Set(key, cacheStore, cacheStore.Cost()) // no expiration time
+		// validate we didn't have a race with another set latest block.
+		// this mechanism could be improved and also changed in the future.
+		go func() {
```
Comment (ranlavanet) [inline ecosystem/cache/handlers.go:223]: @omerlavanet lets think if this is the best solution

File: ecosystem/cache/handlers.go [hunk 3]
```diff
@@ -142,7 +155,11 @@ func (s *RelayerCacheServer) SetRelay(ctx context.Context, relayCacheSet *pairin
 		cache := s.CacheServer.tempCache
 		cache.SetWithTTL(cacheKey, cacheValue, cacheValue.Cost(), s.getExpirationForChain(relayCacheSet.ChainID, relayCacheSet.BlockHash))
 	}
-	s.setLatestBlock(relayCacheSet.ChainID, relayCacheSet.Provider, relayCacheSet.Request.RequestBlock)
+	// Setting the seen block for shared state.
+	// Getting the max block number between the seen block on the consumer side vs the latest block on the response of the provider
+	latestKnownBlock := int64(math.Max(float64(relayCacheSet.Request.RequestBlock), float64(relayCacheSet.Request.SeenBlock)))
```
Comment (ranlavanet) [inline ecosystem/cache/handlers.go:160]: @omerlavanet  provider side and consumer side here might be different with request block vs response block

File: ecosystem/cache/handlers.go [hunk 4]
```diff
@@ -28,7 +29,9 @@ var (
 )
 
 const (
-	SEP = ";"
+	SharedStateWriteAttempts = 5
+	Int64Cost                = 8 // the cost is directly set to 8, which is the size in bytes of the int64
+	SEP                      = ";"
```
Comment (omerlavanet) [inline ecosystem/cache/handlers.go:34]: delete cost

File: ecosystem/cache/handlers.go [hunk 5]
```diff
@@ -66,26 +69,36 @@ func (cv *LastestCacheStore) Cost() int64 {
...
 }
 
-func (s *RelayerCacheServer) getRelayInner(ctx context.Context, relayCacheGet *pairingtypes.RelayCacheGet) (*pairingtypes.CacheRelayReply, error) {
+func (s *RelayerCacheServer) getRelayInner(ctx context.Context, relayCacheGet *pairingtypes.RelayCacheGet, latestKnownBlock int64) (*pairingtypes.CacheRelayReply, error) {
 	inputFormatter, outputFormatter := format.FormatterForRelayRequestAndResponse(relayCacheGet.Request.ApiInterface)
 	relayCacheGet.Request.Data = inputFormatter(relayCacheGet.Request.Data)
 	requestedBlock := relayCacheGet.Request.RequestBlock
-	getLatestBlock := s.getLatestBlock(relayCacheGet.ChainID, relayCacheGet.Provider)
-	relayCacheGet.Request.RequestBlock = lavaprotocol.ReplaceRequestedBlock(requestedBlock, getLatestBlock)
+	relayCacheGet.Request.RequestBlock = lavaprotocol.ReplaceRequestedBlock(requestedBlock, latestKnownBlock)
```
Comment (omerlavanet) [inline ecosystem/cache/handlers.go:101]: let's keep the getLatestBlock original, so the sync and freshness isnt hindered, and separate the seen block and latest block

File: ecosystem/cache/handlers.go [hunk 6]
```diff
@@ -66,26 +69,36 @@ func (cv *LastestCacheStore) Cost() int64 {
...
+			cacheReply = &pairingtypes.CacheRelayReply{}
+		}
 	} else {
 		hit = true
 		s.cacheHit(ctx)
 	}
+
+	// set seen block if required
+	if relayCacheGet.SharedStateId != "" {
+		cacheReply.SeenBlock = latestKnownBlock
```
Comment (omerlavanet) [inline ecosystem/cache/handlers.go:89]: needs to change to sene block after the cache seen block separation

File: ecosystem/cache/handlers.go (hunk 3 above)
Comment (omerlavanet) [reply ecosystem/cache/handlers.go:160]:
```suggestion
	latestKnownBlock := int64(math.Max(float64(relayCacheSet.Response.LatestBlock), float64(relayCacheSet.Request.SeenBlock)))
```

File: ecosystem/cache/handlers.go [hunk 7]
```diff
@@ -185,22 +202,37 @@ func (s *RelayerCacheServer) getLatestBlockInner(chainID string, providerAddr st
...
 		// we are setting this with a futuristic invalidation time, we still want the entry in cache to protect us from putting a lower last block
 		cacheStore := LastestCacheStore{latestBlock: latestBlock, latestExpirationTime: time.Now().Add(DefaultExpirationForNonFinalized)}
-		utils.LavaFormatDebug("setting latest block", utils.Attribute{Key: "providerAddr", Value: providerAddr}, utils.Attribute{Key: "chainID", Value: chainID}, utils.Attribute{Key: "latestBlock", Value: latestBlock})
-		s.CacheServer.finalizedCache.Set(latestBlockKey(chainID, providerAddr), cacheStore, cacheStore.Cost()) // no expiration time
+		utils.LavaFormatDebug("setting latest block", utils.Attribute{Key: "key", Value: key}, utils.Attribute{Key: "latestBlock", Value: latestBlock})
+		s.CacheServer.finalizedCache.Set(key, cacheStore, cacheStore.Cost()) // no expiration time
+		// validate we didn't have a race with another set latest block.
+		// this mechanism could be improved and also changed in the future.
+		go func() {
+			for i := 0; i < SharedStateWriteAttempts; i++ {
```
Comment (omerlavanet) [inline ecosystem/cache/handlers.go:224]: use this mechanism on both latest and seen writes

**Review by omerlavanet — CHANGES_REQUESTED** [review]

**Comment by ranlavanet** [comment]
@omerlavanet PR ready.

File: ecosystem/cache/handlers.go [hunk 8]
```diff
@@ -64,17 +66,42 @@ func (cv *LastestCacheStore) Cost() int64 {
...
+		value, found := getNonExpiredFromCache(s.CacheServer.finalizedCache, id)
+		if !found {
+			utils.LavaFormatInfo("Failed fetching state from cache for this user id", utils.LogAttr("id", id))
+			return 0 // we cant set the seen block in this case it will be returned 0 and wont be used in the consumer side.
+		}
+		utils.LavaFormatInfo("getting seen block cache", utils.LogAttr("id", id), utils.LogAttr("value", value))
+		if cacheValue, ok := value.(int64); ok {
+			return cacheValue
+		}
+		utils.LavaFormatError("Failed converting cache result to int64", nil, utils.LogAttr("value", value))
```
Comment (omerlavanet) [inline ecosystem/cache/handlers.go:81]: should be fatal

File: ecosystem/cache/handlers.go [hunk 9]
```diff
@@ -64,17 +66,42 @@ func (cv *LastestCacheStore) Cost() int64 {
...
+		if cacheReply == nil {
+			cacheReply = &pairingtypes.CacheRelayReply{}
+		}
 	} else {
 		hit = true
 		s.cacheHit(ctx)
 	}
+
+	// set seen block if required
+	cacheReply.SeenBlock = s.getSeenBlockForSharedStateMode(relayCacheGet.ChainID, relayCacheGet.SharedStateId)
```
Comment (omerlavanet) [inline ecosystem/cache/handlers.go:103]: we are doing the two cache reads sequentially doubling the delay, lets do a wait group and do both at the same time

File: ecosystem/cache/handlers.go [hunk 10]
```diff
@@ -120,6 +147,37 @@ func (s *RelayerCacheServer) getRelayInner(ctx context.Context, relayCacheGet *p
...
+	}
+	existingSeenBlock := s.getSeenBlockForSharedStateMode(chainId, sharedStateId)
+	// validate we have a newer block than the existing stored in the db.
+	if existingSeenBlock <= seenBlock { // refreshes state even if its equal
+		key := latestBlockKey(chainId, sharedStateId)
+		// for seen block we expire the entry after one hour otherwise this user will stay in the db for ever
+		s.CacheServer.finalizedCache.SetWithTTL(key, seenBlock, 0, s.CacheServer.ExpirationFinalized)
+		// a validation routine to make sure we don't have a race for seen block rewrites as there are concurrent writes.
+		// this will be solved once we implement a db with a queue but for now its a good enough solution.
+		go func() {
```
Comment (omerlavanet) [inline ecosystem/cache/handlers.go:165]: can we do both retry functions here and on set latest in the same function (get the write as a callback), later we will replace it and we don;t want to do it in many places

File: protocol/chainlib/chain_fetcher.go [hunk 11]
```diff
@@ -93,7 +93,7 @@ func (cf *ChainFetcher) populateCache(relayData *pairingtypes.RelayPrivateData,
 		new_ctx := context.Background()
 		new_ctx, cancel := context.WithTimeout(new_ctx, common.DataReliabilityTimeoutIncrease)
 		defer cancel()
-		err := cf.cache.SetEntry(new_ctx, relayData, requestedBlockHash, cf.endpoint.ChainID, reply, finalized, "", nil)
+		err := cf.cache.SetEntry(new_ctx, &pairingtypes.RelayCacheSet{Request: relayData, BlockHash: requestedBlockHash, ChainID: cf.endpoint.ChainID, Response: reply, Finalized: finalized, OptionalMetadata: nil})
```
Comment (omerlavanet) [inline protocol/chainlib/chain_fetcher.go:96]: let's explain why shared state is "" here and write it explicitly

File: protocol/rpcconsumer/rpcconsumer_server.go [hunk 12]
```diff
@@ -408,12 +411,51 @@ func (rpccs *RPCConsumerServer) sendRelayToProvider(
...
+			// so we can return here, no need to update anything and calculate as this info was fetched from the cache
+			relayResult = &common.RelayResult{Reply: reply, Request: &pairingtypes.RelayRequest{RelayData: relayRequestData}}
+			lavaprotocol.SetRequestedBlockNotApplicable(relayResult.Request.RelayData)
+			return relayResult, nil
+		}
+	} else {
+		utils.LavaFormatDebug("skipping cache due to requested block being NOT_APPLICABLE", utils.Attribute{Key: "api name", Value: chainMessage.GetApi().Name})
+	}
+	// cache failed, move on to regular relay
+	if performance.NotConnectedError.Is(cacheError) {
```
Comment (omerlavanet) [inline protocol/rpcconsumer/rpcconsumer_server.go:455]: why is this outside the if case? it's only applicable if we queried the cache

**Review by omerlavanet — CHANGES_REQUESTED** [review]

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**